### DISCUSSION TITLE
Spam prevention

### DISCUSSION THREAD

**Discussion by aaronc** [regen-network/regen-ledger#255] [Ideas]
What mechanisms should we use?
* consensus minimum fees
* per-operation fees (beyond gas)
* deposits?
* allow/deny lists of permissioned addresses (short-term solution)

Relevant discussions in the SDK: https://github.com/cosmos/cosmos-sdk/discussions/8224 and https://github.com/cosmos/cosmos-sdk/issues/4527

**Comment by clevinson** [comment]
One consideration with deposits is that it probably requires more refactoring of `x/fee_grant` than other approaches (consensus minimum fees or per-operation fees beyond gas).

From an end user standpoint, we definitely want to enable monitors & farmers to be able to send 0-fee transactions, and implementing spam prevention that still allows use of our existing fee-delegation logic seems preferable imo.

After reading into the Cosmos SDK discussions, it seems like there is rough alignment on the following approach from Bez in https://github.com/cosmos/cosmos-sdk/issues/4527#issuecomment-505147344:
> Have some initial consensus min fee F and a target block fulfillment size S (e.g. 75%). Every N blocks check to see if the average block fulfillment size is below or above S. If below, reduce F by some factor. If above, increase F by some factor. Otherwise, leave unchanged. We can probably ignore empty blocks too.

I would say since there's active discussion of this across a few different SDK threads, and since there seems to be general interest in prioritizing this on the SDK side, we should try to see if the above proposed SDK solution meets our needs. If so, let's prioritize for SDK v0.43 and block implementing the new regen modules on mainnet until this is resolved.

That is atleast my preferred approach. Thoughts from others?

**Reply by aaronc** [reply]
I think getting that algorithm into the SDK would be a great start. It doesn't even seem too hard...

**Reply by aaronc** [reply]
I can start on a PR... is anyone else interested in collaborating on this?

**Comment by robert-zaremba** [comment]
I strongly support deposit for data heavy transactions. We could creat interface for that:
* `Deposit()` - deposit fee for 1 Byte of data
* `UserDeposit()` - shows user deposit
* `MakeDepoait(userID | null)`

Storage has inherently differe cost structure than CPU, so i think it make sense to price it differently. Fee grant can have a structure: (max deposit, max fee)
